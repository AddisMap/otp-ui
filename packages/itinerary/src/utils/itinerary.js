export const transitModes = [
  "TRAM",
  "BUS",
  "SUBWAY",
  "FERRY",
  "RAIL",
  "GONDOLA"
];

export function isTransit(mode) {
  return transitModes.includes(mode) || mode === "TRANSIT";
}

// TODO: Is this a place where we can get these colors dynamically?
export function getMapColor(mode) {
  switch (mode) {
    case "WALK":
      return "#444";
    case "BICYCLE":
      return "#0073e5";
    case "SUBWAY":
      return "#f00";
    case "RAIL":
      return "#b00";
    case "BUS":
      return "#080";
    case "TRAM":
      return "#800";
    case "FERRY":
      return "#008";
    case "CAR":
      return "#444";
    case "MICROMOBILITY":
      return "#f5a729";
    default:
      return "#aaa";
  }
}

// TODO: temporary code; handle via migrated OTP i18n language table
export function getStepDirection(step) {
  switch (step.relativeDirection) {
    case "DEPART":
      return `Head ${step.absoluteDirection.toLowerCase()}`;
    case "LEFT":
      return "Left";
    case "HARD_LEFT":
      return "Hard left";
    case "SLIGHTLY_LEFT":
      return "Slight left";
    case "CONTINUE":
      return "Continue";
    case "SLIGHTLY_RIGHT":
      return "Slight right";
    case "RIGHT":
      return "Right";
    case "HARD_RIGHT":
      return "Hard right";
    case "CIRCLE_CLOCKWISE":
      return "Follow circle clockwise";
    case "CIRCLE_COUNTERCLOCKWISE":
      return "Follow circle counterclockwise";
    case "ELEVATOR":
      return "Take elevator";
    case "UTURN_LEFT":
      return "Left U-turn";
    case "UTURN_RIGHT":
      return "Right U-turn";
    default:
      return step.relativeDirection;
  }
}

export function getStepInstructions(step) {
  const conjunction = step.relativeDirection === "ELEVATOR" ? "to" : "on";
  return `${getStepDirection(step)} ${conjunction} ${step.streetName}`;
}

export function getStepStreetName(step) {
  if (step.streetName === "road") return "Unnamed Road";
  if (step.streetName === "path") return "Unnamed Path";
  return step.streetName;
}

/**
 * Get the configured company object for the given network string if the company
 * has been defined in the provided companies array config.
 */
function getCompanyForNetwork(networkString, companies = []) {
  const company = companies.find(co => co.id === networkString);
  if (!company) {
    console.warn(
      `No company found in config.yml that matches rented vehicle network: ${networkString}`,
      companies
    );
  }
  return company;
}

/**
 * Returns mode name by checking the vertex type (VertexType class in OTP) for
 * the provided place. NOTE: this is currently only intended for vehicles at
 * the moment (not transit or walking).
 *
 * TODO: I18N
 * @param  {string} place place from itinerary leg
 */
export function getModeForPlace(place) {
  switch (place.vertexType) {
    case "CARSHARE":
      return "car";
    case "VEHICLERENTAL":
      return "E-scooter";
    // TODO: Should the type change depending on bike vertex type?
    case "BIKESHARE":
    case "BIKEPARK":
      return "bike";
    // If company offers more than one mode, default to `vehicle` string.
    default:
      return "vehicle";
  }
}

export function getPlaceName(place, companies) {
  // If address is provided (i.e. for carshare station, use it)
  if (place.address) return place.address.split(",")[0];
  if (place.networks && place.vertexType === "VEHICLERENTAL") {
    // For vehicle rental pick up, do not use the place name. Rather, use
    // company name + vehicle type (e.g., SPIN E-scooter). Place name is often just
    // a UUID that has no relevance to the actual vehicle. For bikeshare, however,
    // there are often hubs or bikes that have relevant names to the user.
    const company = getCompanyForNetwork(place.networks[0], companies);
    if (company) {
      return `${company.label} ${getModeForPlace(place)}`;
    }
  }
  // Default to place name
  return place.name;
}

export function toSentenceCase(str) {
  if (str == null) {
    return "";
  }
  const typedString = String(str);
  return (
    typedString.charAt(0).toUpperCase() + typedString.substr(1).toLowerCase()
  );
}

export function getLegModeLabel(leg) {
  switch (leg.mode) {
    case "BICYCLE_RENT":
      return "Biketown";
    case "CAR":
      return leg.hailedCar ? "Ride" : "Drive";
    case "GONDOLA":
      return "Aerial Tram";
    case "TRAM":
      if (leg.routeLongName.toLowerCase().indexOf("streetcar") !== -1)
        return "Streetcar";
      return "Light Rail";
    case "MICROMOBILITY":
      return "Ride";
    default:
      return toSentenceCase(leg.mode);
  }
}

/** Looks up an operator from the provided configuration */
export const getOperatorFromConfig = (id, config) =>
  config.operators.find(operator => operator.id === id) || null;
