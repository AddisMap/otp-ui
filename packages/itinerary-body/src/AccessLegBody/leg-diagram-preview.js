import { getElevationProfile } from "@opentripplanner/core-utils/lib/itinerary";
import { legType } from "@opentripplanner/core-utils/lib/types";
import PropTypes from "prop-types";
import React, { Component } from "react";
import ReactResizeDetector from "react-resize-detector";

const METERS_TO_FEET = 3.28084;

function generateSvg(profile, width) {
  const height = 30;
  const { points: ptArr, traversed } = profile;
  let { minElev, maxElev } = profile;
  // Pad the min-max range by 25m on either side
  minElev -= 25;
  maxElev += 25;

  // Transform the point array and store it as an SVG-ready string
  const pts = ptArr
    .map(pt => {
      const x = (pt[0] / traversed) * width;
      const y = height - (height * (pt[1] - minElev)) / (maxElev - minElev);
      return `${x},${y}`;
    })
    .join(" ");

  // Render the SVG
  return (
    <svg height={height} width={width}>
      <polyline points={pts} fill="none" stroke="black" strokeWidth={1.3} />
    </svg>
  );
}

class LegDiagramPreview extends Component {
  constructor(props) {
    super(props);
    this.state = { width: null };
  }

  onResize = width => {
    if (width > 0) {
      this.setState({ width });
    }
  };

  /**
   * Determine if the diagram currently visible is for this leg (based on start
   * time).
   */
  isActive = () => {
    const { diagramVisible, leg } = this.props;
    return diagramVisible && diagramVisible.startTime === leg.startTime;
  };

  onExpandClick = () => {
    const { leg, showLegDiagram } = this.props;
    if (this.isActive()) showLegDiagram(null);
    else showLegDiagram(leg);
  };

  /** Round elevation to whole number and add symbol. */
  formatElevation = elev => `${Math.round(elev)}'`;

  render() {
    const { leg, showElevationProfile } = this.props;
    const { width } = this.state;
    if (!showElevationProfile) return null;
    const profile = getElevationProfile(leg.steps);
    // Don't show for very short legs
    if (leg.distance < 500 || leg.mode === "CAR") return null;

    return (
      <div className={`leg-diagram-preview ${this.isActive() ? "on" : ""}`}>
        {/* The preview elevation SVG */}
        {/* eslint-disable-next-line */}
        <div
          className="diagram"
          tabIndex="0"
          title="Toggle elevation chart"
          role="button"
          onClick={this.onExpandClick}
        >
          <div className="diagram-title text-center">
            Elevation chart{" "}
            <span style={{ fontSize: "xx-small", color: "red" }}>
              ↑{this.formatElevation(profile.gain * METERS_TO_FEET)}
              {"  "}
            </span>
            <span style={{ fontSize: "xx-small", color: "green" }}>
              ↓{this.formatElevation(-profile.loss * METERS_TO_FEET)}
            </span>
          </div>
          {profile.points.length > 0
            ? generateSvg(profile, width)
            : "No elevation data available."}
          <ReactResizeDetector handleWidth onResize={this.onResize} />
        </div>
      </div>
    );
  }
}

LegDiagramPreview.propTypes = {
  diagramVisible: PropTypes.bool.isRequired,
  leg: legType.isRequired,
  showElevationProfile: PropTypes.bool.isRequired,
  showLegDiagram: PropTypes.func.isRequired
};

export default LegDiagramPreview;
