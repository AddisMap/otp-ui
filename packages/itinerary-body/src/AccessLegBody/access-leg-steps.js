import { DirectionIcon } from "@opentripplanner/icons";
import React, { ReactElement } from "react";
import { FormattedMessage } from "react-intl";
import { Step } from "@opentripplanner/types";

import * as S from "../styled";
import MapillaryButton from "./mapillary-button";

interface Props {
  mapillaryCallback?: (id: string) => void;
  mapillaryKey?: string;
  steps: Step[];
}

/**
 * Applies the appropriate style for street names.
 */
function renderStreetName(streetName) {
  return <S.StepStreetName>{streetName}</S.StepStreetName>;
}

/**
 * Renders a turn-by-turn step of an access leg.
 */
export default function AccessLegSteps({
  steps,
  mapillaryCallback,
  mapillaryKey
}: Props): ReactElement {
  return (
    <S.Steps>
      {steps.map(
        ({ absoluteDirection, lat, lon, relativeDirection, streetName }, k) => (
          <S.StepRow key={k}>
            <S.StepIconContainer>
              <DirectionIcon relativeDirection={relativeDirection} />
            </S.StepIconContainer>

            <S.StepDescriptionContainer>
              <FormattedMessage
                defaultMessage="{relativeDirection} {absoluteDirection} on <strong>{streetName}</strong>"
                description="Describes a step of the directions to reach a destination."
                id="otpUi.AccessLegBody.step"
                values={{
                  absoluteDirection,
                  relativeDirection,
                  streetName,
                  strong: renderStreetName
                }}
              />

              <MapillaryButton
                clickCallback={mapillaryCallback}
                coords={{ lat, lon }}
                mapillaryKey={mapillaryKey}
                padLeft
              />
            </S.StepDescriptionContainer>
          </S.StepRow>
        )
      )}
    </S.Steps>
  );
}
