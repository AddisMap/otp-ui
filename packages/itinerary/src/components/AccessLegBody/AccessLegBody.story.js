import React from "react";
import { storiesOf } from "@storybook/react";
import { withA11y } from "@storybook/addon-a11y";
import { withInfo } from "@storybook/addon-info";
import { action } from "@storybook/addon-actions";
import { withThemesProvider } from "storybook-addon-styled-component-theme";
import AccessLegBody from ".";
import itinerary from "../../__mocks__/itinerary";
import ThemeShim from "../../common/components/ThemeShim";
import MainStyleDecorator from "../../common/components/MainStyleDecorator";
import {
  LIGHT_THEME,
  DARK_THEME
} from "../../common/constants";
import config from "../../__mocks__/config";

const customIcons = [];
const followsTransit = false;
const leg = itinerary.legs[0];
const timeOptions = { format: "h:mm a" };
const legIndex = 0;
const routingType = "ITINERARY";
const setActiveLeg = action("setActiveLeg");

storiesOf("Itinerary|Access Leg Body", module)
  // .addDecorator(withThemesProvider([LIGHT_THEME, DARK_THEME]))
  .addDecorator(withA11y)
  .addDecorator(withInfo)
  .addDecorator(MainStyleDecorator)
  .add("Access Leg Body", () => (
    <ThemeShim>
      <AccessLegBody
        config={config}
        leg={leg}
        customIcons={customIcons}
        legIndex={legIndex}
        timeOptions={timeOptions}
        followsTransit={followsTransit}
        setActiveLeg={setActiveLeg}
        routingType={routingType}
      />
    </ThemeShim>
  ));
