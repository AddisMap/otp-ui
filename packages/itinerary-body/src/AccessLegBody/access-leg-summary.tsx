import coreUtils from "@opentripplanner/core-utils";
import { humanizeDistanceString } from "@opentripplanner/humanize-distance";
import { Config, Leg } from "@opentripplanner/types";
import React, { FunctionComponent, ReactElement } from "react";
import { FormattedMessage, useIntl } from "react-intl";

import * as Styled from "../styled";

interface Props {
  config: Config;
  leg: Leg;
  LegIcon: FunctionComponent<{ leg: Leg }>;
  onSummaryClick: () => void;
  showLegIcon: boolean;
}

export default function AccessLegSummary({
  config,
  leg,
  LegIcon,
  onSummaryClick,
  showLegIcon
}: Props): ReactElement {
  const intl = useIntl();
  return (
    <Styled.LegClickable onClick={onSummaryClick}>
      {showLegIcon && (
        <Styled.LegIconContainer>
          <LegIcon leg={leg} />
        </Styled.LegIconContainer>
      )}

      {/* Leg description, e.g. "Walk 0.5 mi to..." */}
      <Styled.LegDescription>
        <FormattedMessage
          defaultMessage="{modeId} {distance} to {place}"
          description="Summarizes an access leg"
          id="otpUi.AccessLegBody.summary"
          values={{
            // TODO: Implement metric vs imperial (up until now it's just imperial).
            distance:
              leg.distance > 0
                ? humanizeDistanceString(leg.distance, false, intl)
                : 0,
            isCarHail: leg.hailedCar,
            modeId: leg.mode,
            // FIXME: localize rental device names.
            place: coreUtils.itinerary.getPlaceName(leg.to, config.companies)
          }}
        />
      </Styled.LegDescription>
    </Styled.LegClickable>
  );
}
