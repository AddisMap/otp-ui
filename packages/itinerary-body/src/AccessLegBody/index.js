import currencyFormatter from "currency-formatter";
import {
  getLegModeLabel,
  getPlaceName,
  getStepDirection,
  getStepStreetName
} from "@opentripplanner/core-utils/lib/itinerary";
import {
  formatDuration,
  formatTime
} from "@opentripplanner/core-utils/lib/time";
import {
  configType,
  legType,
  stepsType,
  timeOptionsType
} from "@opentripplanner/core-utils/lib/types";
import { isMobile } from "@opentripplanner/core-utils/lib/ui";
import { humanizeDistanceString } from "@opentripplanner/humanize-distance";
import { DirectionIcon } from "@opentripplanner/icons/lib/directions";
import React, { Component } from "react";
import PropTypes from "prop-types";
import { VelocityTransitionGroup } from "velocity-react";

import LegDiagramPreview from "./leg-diagram-preview";
import * as Styled from "../styled";

/**
 * Component for access (e.g. walk/bike/etc.) leg in narrative itinerary. This
 * particular component is used in the line-itin (i.e., trimet-mod-otp) version
 * of the narrative itinerary.
 */
export default class AccessLegBody extends Component {
  constructor(props) {
    super(props);
    this.state = { expanded: false };
  }

  onStepsHeaderClick = () => {
    const { expanded } = this.state;
    this.setState({ expanded: !expanded });
  };

  onSummaryClick = () => {
    const { leg, legIndex, setActiveLeg } = this.props;
    setActiveLeg(legIndex, leg);
  };

  render() {
    const {
      config,
      diagramVisible,
      followsTransit,
      leg,
      LegIcon,
      routingType,
      setLegDiagram,
      showElevationProfile,
      timeOptions
    } = this.props;
    const { expanded } = this.state;

    if (leg.mode === "CAR" && leg.hailedCar) {
      return (
        <TNCLeg
          config={config}
          leg={leg}
          LegIcon={LegIcon}
          onSummaryClick={this.onSummaryClick}
          timeOptions={timeOptions}
          followsTransit={followsTransit}
        />
      );
    }

    return (
      <Styled.LegBody>
        <AccessLegSummary
          config={config}
          leg={leg}
          LegIcon={LegIcon}
          onSummaryClick={this.onSummaryClick}
        />
        <Styled.StepsHeader onClick={this.onStepsHeaderClick}>
          {formatDuration(leg.duration)}
          {leg.steps && (
            <span>
              {" "}
              <Styled.CaretToggle expanded />
            </span>
          )}
        </Styled.StepsHeader>

        {routingType === "ITINERARY" && (
          <LegDiagramPreview
            diagramVisible={diagramVisible}
            leg={leg}
            setLegDiagram={setLegDiagram}
            showElevationProfile={showElevationProfile}
          />
        )}
        <VelocityTransitionGroup
          enter={{ animation: "slideDown" }}
          leave={{ animation: "slideUp" }}
        >
          {expanded && <AccessLegSteps steps={leg.steps} />}
        </VelocityTransitionGroup>
      </Styled.LegBody>
    );
  }
}

AccessLegBody.propTypes = {
  config: configType.isRequired,
  /**
   * Should be either null or a legType. Indicates that a particular leg diagram
   * has been selected and is active.
   */
  diagramVisible: legType,
  followsTransit: PropTypes.bool,
  leg: legType.isRequired,
  LegIcon: PropTypes.elementType.isRequired,
  legIndex: PropTypes.number.isRequired,
  routingType: PropTypes.string.isRequired,
  setActiveLeg: PropTypes.func.isRequired,
  setLegDiagram: PropTypes.func.isRequired,
  showElevationProfile: PropTypes.bool,
  timeOptions: timeOptionsType
};

AccessLegBody.defaultProps = {
  diagramVisible: null,
  followsTransit: false,
  showElevationProfile: false,
  timeOptions: null
};

function TNCLeg({
  config,
  LYFT_CLIENT_ID,
  UBER_CLIENT_ID,
  followsTransit,
  leg,
  LegIcon,
  onSummaryClick,
  timeOptions
}) {
  const universalLinks = {
    UBER: `https://m.uber.com/${
      isMobile() ? "ul/" : ""
    }?client_id=${UBER_CLIENT_ID}&action=setPickup&pickup[latitude]=${
      leg.from.lat
    }&pickup[longitude]=${leg.from.lon}&pickup[formatted_address]=${encodeURI(
      leg.from.name
    )}&dropoff[latitude]=${leg.to.lat}&dropoff[longitude]=${
      leg.to.lon
    }&dropoff[formatted_address]=${encodeURI(leg.to.name)}`,
    LYFT: `https://lyft.com/ride?id=lyft&partner=${LYFT_CLIENT_ID}&pickup[latitude]=${leg.from.lat}&pickup[longitude]=${leg.from.lon}&destination[latitude]=${leg.to.lat}&destination[longitude]=${leg.to.lon}`
  };
  const { tncData } = leg;

  if (!tncData || !tncData.estimatedArrival) return null;
  return (
    <div>
      <Styled.PlaceSubheader>
        Wait{" "}
        {!followsTransit && (
          <span>{Math.round(tncData.estimatedArrival / 60)} minutes </span>
        )}
        for {tncData.displayName} pickup
      </Styled.PlaceSubheader>

      <Styled.LegBody>
        {/* The icon/summary row */}
        <AccessLegSummary
          config={config}
          leg={leg}
          LegIcon={LegIcon}
          onSummaryClick={onSummaryClick}
        />

        {/* The "Book Ride" button */}
        <Styled.BookTNCRideButtonContainer>
          <Styled.BookTNCRideButton
            href={universalLinks[tncData.company]}
            target={isMobile() ? "_self" : "_blank"}
          >
            Book Ride
          </Styled.BookTNCRideButton>
          {followsTransit && <Styled.BookLaterPointer />}
          {followsTransit && (
            <Styled.BookLaterContainer>
              <Styled.BookLaterInnerContainer>
                <Styled.BookLaterText>
                  Wait until{" "}
                  {formatTime(
                    leg.startTime - tncData.estimatedArrival * 1000,
                    timeOptions
                  )}{" "}
                  to book
                </Styled.BookLaterText>
              </Styled.BookLaterInnerContainer>
            </Styled.BookLaterContainer>
          )}
        </Styled.BookTNCRideButtonContainer>

        {/* The estimated travel time */}
        <Styled.StepsHeader>
          Estimated travel time: {formatDuration(leg.duration)} (does not
          account for traffic)
        </Styled.StepsHeader>

        {/* The estimated travel cost */}
        {tncData.minCost && (
          <p>
            Estimated cost:{" "}
            {`${currencyFormatter.format(tncData.minCost, {
              code: tncData.currency
            })} - ${currencyFormatter.format(tncData.maxCost, {
              code: tncData.currency
            })}`}
          </p>
        )}
      </Styled.LegBody>
    </div>
  );
}

TNCLeg.propTypes = {
  config: configType.isRequired,
  LYFT_CLIENT_ID: PropTypes.string,
  UBER_CLIENT_ID: PropTypes.string,
  followsTransit: PropTypes.bool.isRequired,
  leg: legType.isRequired,
  LegIcon: PropTypes.elementType.isRequired,
  onSummaryClick: PropTypes.func.isRequired,
  timeOptions: timeOptionsType
};

TNCLeg.defaultProps = {
  LYFT_CLIENT_ID: "",
  UBER_CLIENT_ID: "",
  timeOptions: null
};

function AccessLegSummary({ config, leg, /* LegIcon, */ onSummaryClick }) {
  return (
    <Styled.LegClickable onClick={onSummaryClick}>
      {/* Mode-specific icon
      <div>
        <Styled.LegIconContainer>
          <LegIcon leg={leg} />
        </Styled.LegIconContainer>
      </div> */}

      {/* Leg description, e.g. "Walk 0.5 mi to..." */}
      <div>
        {getLegModeLabel(leg)}{" "}
        {leg.distance > 0 && (
          <span> {humanizeDistanceString(leg.distance)}</span>
        )}
        {` to ${getPlaceName(leg.to, config.companies)}`}
      </div>
    </Styled.LegClickable>
  );
}

AccessLegSummary.propTypes = {
  config: configType.isRequired,
  leg: legType.isRequired,
  // LegIcon: PropTypes.elementType.isRequired,
  onSummaryClick: PropTypes.func.isRequired
};

function AccessLegSteps({ steps }) {
  return (
    <Styled.Steps>
      {steps.map((step, k) => {
        return (
          <Styled.StepRow key={k}>
            <Styled.StepIconContainer>
              <DirectionIcon relativeDirection={step.relativeDirection} />
            </Styled.StepIconContainer>

            <Styled.StepDescriptionContainer>
              {getStepDirection(step)}
              <span>
                {step.relativeDirection === "ELEVATOR" ? " to " : " on "}
              </span>
              <Styled.StepStreetName>
                {getStepStreetName(step)}
              </Styled.StepStreetName>
            </Styled.StepDescriptionContainer>
          </Styled.StepRow>
        );
      })}
    </Styled.Steps>
  );
}

AccessLegSteps.propTypes = {
  steps: stepsType.isRequired
};
