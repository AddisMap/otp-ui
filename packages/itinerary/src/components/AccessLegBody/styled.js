import styled from "styled-components";
import { ClearButton } from "../../common/components/Styled";
import BaseExpandIcon from "../ExpandIcon";
import BaseDirectionIcon from "../DirectionIcon";
import { toModeBorderColor } from "../../utils/mode";

export const LegBody = styled.div`
  padding: 15px 0 15px 15px;
`;

export const StepsHeader = styled(ClearButton).attrs(props => ({
  "aria-pressed": props.expanded
}))`
  padding: 5px;
`;

export const ExpandIcon = styled(BaseExpandIcon).attrs(props => ({
  width: 10,
  height: 10,
  fill: toModeBorderColor(props.mode)
}))`
  padding-left: 5px;
`;

export const AccessLegSummary = styled(ClearButton).attrs(props => ({
  "aria-pressed": props.expanded
}))`
  width: 100%;
  font-size: 1.1em;
`;

export const AccessStepsListWrapper = styled.div`
  overflow: hidden;
  height: auto;
  max-height: ${props => (props.expanded === "true" ? "1000px" : "0")};
  transition: max-height 1s ease-in-out;
`;

export const AccessStepsList = styled.div`
  padding: 5px;
`;

export const AccessStepRow = styled.div`
  flex-flow: row;
  display: flex;
`;

export const StepDirectionIconColumn = styled.div`
  flex: 0 0 20px;
`;

export const StepDirectionDetailsColumn = styled.div`
  flex: 1 1 auto;
`;

export const StepDirectionIcon = styled(BaseDirectionIcon).attrs(props => ({
  width: 15,
  height: 15,
  fill: props.theme.primaryFontColor
}))``;

export const StepStreetName = styled.span`
  font-weight: bold;
`;
