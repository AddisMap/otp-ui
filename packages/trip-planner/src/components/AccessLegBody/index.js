import React, { useState } from "react";
import PropTypes from "prop-types";
import * as Styled from "./styled";
import { formatDuration } from "../../utils/time";
import * as Accessibility from "~/packages/trip-planner/src/common/components/Accessibility";
import {
  getLegModeLabel,
  getPlaceName,
  getStepDirection,
  getStepStreetName
} from "../../utils/itinerary";
import { distanceString } from "../../utils/distance";

/*
 TODO: Since we're still working on this component, 
 let's retain the original set of props to match OTP-RR
*/
const AccessLegBody = ({
  config,
  // eslint-disable-next-line no-unused-vars
  customIcons,
  // eslint-disable-next-line no-unused-vars
  followsTransit,
  leg,
  // eslint-disable-next-line no-unused-vars
  timeOptions,
  // eslint-disable-next-line no-unused-vars
  legIndex,
  // eslint-disable-next-line no-unused-vars
  routingType,
  // eslint-disable-next-line no-unused-vars
  setActiveLeg
}) => {
  const [expanded, setExpanded] = useState(false);

  return (
    <Styled.LegBody>
      {leg.mode === "CAR" && leg.hailedCar && (
        <div>TODO: TNC Leg goes here</div>
      )
      // TODO: TNC leg, what is this. Need to be incorporated
      // <TNCLeg
      //   config={config}
      //   leg={leg}
      //   onSummaryClick={this._onSummaryClick}
      //   timeOptions={timeOptions}
      //   followsTransit={followsTransit}
      //   customIcons={customIcons} />
      }

      <Styled.AccessLegSummary
        onClick={() => setExpanded(!expanded)}
        expanded={expanded}
      >
        {/* Leg description, e.g. "Walk 0.5 mi to..." */}
        {getLegModeLabel(leg)}{" "}
        {leg.distance > 0 && <> {distanceString(leg.distance)}</>}
        {` to ${getPlaceName(leg.to, config.companies)}`}
      </Styled.AccessLegSummary>

      <Styled.StepsHeader
        onClick={() => setExpanded(!expanded)}
        expanded={expanded.toString()}
      >
        <Accessibility.SROnly>
          Click to expand steps for this leg.
        </Accessibility.SROnly>
        {/* TODO: Duration should say 'minute' */}
        {formatDuration(leg.duration)}
        {leg.mode === "WALK" ? " walk" : " ride"}

        {leg.steps && (
          <Styled.ExpandIcon expanded={expanded.toString()} mode={leg.mode} />
        )}
      </Styled.StepsHeader>

      {/* TODO: Leg Diagram preview component -- shows elevation */}
      {/* {routingType === 'ITINERARY' && (
        <LegDiagramPreview leg={leg} />
      )} */}

      <Styled.AccessStepsListWrapper expanded={expanded.toString()}>
        {expanded && (
          <Styled.AccessStepsList>
            {leg.steps.map((step, i) => (
              // eslint-disable-next-line react/no-array-index-key
              <Styled.AccessStepRow key={i}>
                <Styled.StepDirectionIconColumn>
                  <Accessibility.SRHidden>
                    <Styled.StepDirectionIcon
                      direction={step.relativeDirection}
                    />
                  </Accessibility.SRHidden>
                </Styled.StepDirectionIconColumn>
                <Styled.StepDirectionDetailsColumn>
                  {getStepDirection(step)}
                  {step.relativeDirection === "ELEVATOR" ? " to " : " on "}
                  <Styled.StepStreetName>
                    {getStepStreetName(step)}
                  </Styled.StepStreetName>
                </Styled.StepDirectionDetailsColumn>
              </Styled.AccessStepRow>
            ))}
          </Styled.AccessStepsList>
        )}
      </Styled.AccessStepsListWrapper>
    </Styled.LegBody>
  );
};

AccessLegBody.propTypes = {
  /** TODO: Routing Type is usually 'ITINERARY' but we should get more details on what this does */
  routingType: PropTypes.string.isRequired,
  /** Contains details about leg object that is being displayed */
  leg: PropTypes.shape({
    mode: PropTypes.string.isRequired,
    hailedCar: PropTypes.bool.isRequired,
    distance: PropTypes.number.isRequired,
    to: PropTypes.shape({}).isRequired,
    duration: PropTypes.number.isRequired,
    steps: PropTypes.arrayOf(PropTypes.shape({}).isRequired).isRequired
  }).isRequired,
  /** OTP configuration object */
  config: PropTypes.shape({
    companies: PropTypes.arrayOf(PropTypes.shape({})).isRequired
  }).isRequired,
  /** Custom icons that may be used to override icons included in this library */
  customIcons: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  /** The index value of this specific leg within the itinerary */
  legIndex: PropTypes.number.isRequired,
  /** Contains the preferred format string for time display -- may be able to get this from config */
  timeOptions: PropTypes.shape({}).isRequired,
  /** Indicates whether this leg directly follows a transit leg */
  followsTransit: PropTypes.bool.isRequired,
  /** Sets the active leg */
  setActiveLeg: PropTypes.func.isRequired
};

export default AccessLegBody;
