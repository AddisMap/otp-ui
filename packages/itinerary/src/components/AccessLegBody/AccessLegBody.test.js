import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import MOCK_THEME from "~/packages/trip-planner/src/common/__mocks__/theme";
import AccessLegBody from ".";
import itinerary from "../../__mocks__/itinerary";
import config from "../../__mocks__/config";

describe("AccessLegBody Component", () => {
  let sut;

  const customIcons = [];
  const followsTransit = false;
  const leg = itinerary.legs[0];
  const timeOptions = { format: "h:mm a" };
  const legIndex = 0;
  const routingType = "ITINERARY";
  const setActiveLeg = jest.fn(() => {});

  beforeEach(() => {
    sut = render(
      <ThemeProvider theme={MOCK_THEME}>
        <AccessLegBody
          config={config}
          leg={leg}
          customIcons={customIcons}
          legIndex={legIndex}
          timeOptions={timeOptions}
          followsTransit={followsTransit}
          setActiveLeg={setActiveLeg}
          routingType={routingType}
        />
      </ThemeProvider>
    );
  });

  afterEach(() => jest.clearAllMocks());

  it("should match the snapshot", () => {
    const { asFragment } = sut;
    expect(asFragment()).toMatchSnapshot();
  });
});
