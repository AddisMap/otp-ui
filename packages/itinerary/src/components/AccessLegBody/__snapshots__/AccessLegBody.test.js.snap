// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessLegBody Component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bwzfXH kGwkzX"
  >
    <button
      aria-pressed="false"
      class="sc-bdVaJa sc-ifAKCX hSXSQR"
    >
      Walk  0.3 miles to NE Sandy & Cesar Chavez Blvd
    </button>
    <button
      aria-pressed="false"
      class="sc-bdVaJa sc-htpNat ciRYoc"
    >
      <span
        class="sc-gqjmRU kbyrUV"
      >
        Click to expand steps for this leg.
      </span>
      6 min walk
      <svg
        class="sc-bxivhb IpLcM"
        expanded="false"
        fill="#484848"
        height="10"
        mode="WALK"
        viewBox="0 0 24 24"
        width="10"
      >
        <path
          d="M0 7.33L2.829 4.5l9.175 9.339L21.171 4.5 24 7.33 12.004 19.5z"
        />
      </svg>
    </button>
    <div
      class="sc-EHOje fpufHo"
    />
  </div>
</DocumentFragment>
`;
